### DISCUSSION TITLE
Duplicate IKE SA between two hosts

### DISCUSSION THREAD

**Discussion by Jiawei0227** [strongswan/strongswan#2003] [Q&A]
Hey strongswan experts! 

# Setup and my intention
```
Host 1        <----->      Host 2
10.200.0.1                 10.200.0.2
udp/any         ----->     udp/geneve
udp/geneve      <-----     udp/any
```

I am trying to create a setup where certain traffic(Geneve traffic) between two hosts are IPSec encrypted. Other parts etc are skipped(I am able to get it done with local_ts and remote_ts). And I am trying to set it up with transport mode. Here is my configuration and I am using strongswan `5.9.10`

Host 1 which is ip 10.200.0.1
```
connections {
  node2node {
    children {

      send {
        esp_proposals = aes256gcm16
        mode = transport
        start_action = trap
        local_ts = 10.200.0.1/32[udp/any]
        remote_ts = 10.200.0.2/32[udp/6081]
      }

      receive {
        esp_proposals = aes256gcm16
        mode = transport
        start_action = none
        local_ts = 10.200.0.1/32[udp/6081]
        remote_ts = 10.200.0.2/32[udp/any]
      }

    }

    encap = yes
    keyingtries = 1

    # Use IKEv2
    version = 2

    local_addrs = 10.200.0.1/32
    remote_addrs = 10.200.0.2/32
    rekey_time = 3600

    # Encryption algo that used during IKE protocol
    proposals = aes256-sha384-ecp384
    local {
      auth = psk
      id = 10.200.0.1/32
    }
    remote {
      auth = psk
      id = 10.200.0.2/32
    }
  }
}
secrets {
  ike-node2node {
    secret = "foobar"
    id = 10.200.0.2/32
  }
}
```

Host 2 which is ip 10.200.0.2
```
connections {
  node2node {
    children {

      send {
        esp_proposals = aes256gcm16
        mode = transport
        start_action = trap
        local_ts = 10.200.0.2/32[udp/any]
        remote_ts = 10.200.0.1/32[udp/6081]
      }

      receive {
        esp_proposals = aes256gcm16
        mode = transport
        start_action = none
        local_ts = 10.200.0.2/32[udp/6081]
        remote_ts = 10.200.0.1/32[udp/any]
      }

    }

    encap = yes
    keyingtries = 1

    # Use IKEv2
    version = 2

    local_addrs = 10.200.0.2/32
    remote_addrs = 10.200.0.1/32
    rekey_time = 3600

    # Encryption algo that used during IKE protocol
    proposals = aes256-sha384-ecp384
    local {
      auth = psk
      id = 10.200.0.2/32
    }
    remote {
      auth = psk
      id = 10.200.0.1/32
    }
  }
}
secrets {
  ike-node2node {
    secret = "foobar"
    id = 10.200.0.1/32
  }
}
```

With the above configuration. I expect only 1 IKE SA getting created. 2 CHILD_SA pairs creates under the IKE SA. However, what I got is always:

```
node2node: #2, ESTABLISHED, IKEv2, 2fb0677b5a9d92a2_i* 0ea00e6fb323bb1c_r
  local  '10.200.0.2' @ 10.200.0.2[4500]
  remote '10.200.0.1' @ 10.200.0.1[4500]
  AES_CBC-256/HMAC_SHA2_384_192/PRF_HMAC_SHA2_384/ECP_384
  established 901s ago, rekeying in 2600s
  send: #3, reqid 1, INSTALLED, TRANSPORT-in-UDP, ESP:AES_GCM_16-256
    installed 901s ago, rekeying in 2661s, expires in 3059s
    in  cf5afbed,      0 bytes,     0 packets
    out ca430556, 1996749407 bytes, 2787478 packets,     0s ago
    local  10.200.0.2/32[udp]
    remote 10.200.0.1/32[udp/geneve]
node2node: #1, ESTABLISHED, IKEv2, ea58ebcc6fa047fe_i f58f2f963dafbfb7_r*
  local  '10.200.0.2' @ 10.200.0.2[4500]
  remote '10.200.0.1' @ 10.200.0.1[4500]
  AES_CBC-256/HMAC_SHA2_384_192/PRF_HMAC_SHA2_384/ECP_384
  established 1004s ago, rekeying in 2592s
  receive: #2, reqid 2, INSTALLED, TRANSPORT-in-UDP, ESP:AES_GCM_16-256
    installed 1004s ago, rekeying in 2394s, expires in 2956s
    in  c3c5f54a, 3206789455 bytes, 3578330 packets,     0s ago
    out c5e198da,      0 bytes,     0 packets
    local  10.200.0.2/32[udp/geneve]
    remote 10.200.0.1/32[udp]
```

# Question
My understanding is that it suggests two IKE SA are created. And under each IKE SA there is one CHILD_SA pair. I couldnt figure out why because I was under impression the second IKE SA won't be created because there is already an IKA SA initialized by the other end to use. 

- I have ruled out race condition by starting one side and swanctl --load-conns etc to ensure two connections are not created at the same time. So far nothing really works.
- I have tried to setup unique to `keep` but it still shows two IKE SA(which I was under impression it wont create duplicate IKE SA?
- I have read through examples of IKA_SAs but there seems no host-to-host encryption under transport mode like the setup I have
- It appears to work if I set both `start_action` to `none` and use `swanctl --initiate` to start both connection. Then it looks nice and there is only one IKE SA created. 

Can anyone help me to understand where am I configuring it wrong or this is actually expected(two IKE SA between two hosts) under policy trap?

**Comment by Jiawei0227** [comment]
One interesting issue I have found online https://wiki.strongswan.org/issues/2958 which mentioned by @tobiasbrunner 

> The problem is that whenever an acquire is received from the kernel when traffic matches an outbound trap policy, a new IKE_SA is forcibly created (the usual mechanism would otherwise return preexisting IKE_SAs with other peers/IPs).

What I cannot reason about is why in my case there is a new IKE_SA created instead of the existing IKE_SA getting replaced because its the same identity between two hosts.

**Comment by tobiasbrunner** [comment]
> * I have ruled out race condition by starting one side and swanctl --load-conns etc to ensure two connections are not created at the same time. So far nothing really works.

I don't know that you mean here, because as soon as you load the connection (which you need to respond to the peer) that has a child section with `start_action=trap`, traffic can trigger an IKE_SA to get initiated on each of the peers concurrently. And you see that in the status output (the `*` behind the SPIs), each IKE_SA is initiated by one of them.

> I have tried to setup unique to `keep` but it still shows two IKE SA(which I was under impression it wont create duplicate IKE SA?

The uniqueness check won't help if the SAs are created pretty much concurrently.

> Can anyone help me to understand where am I configuring it wrong or this is actually expected(two IKE SA between two hosts) under policy trap?

If you configure trap policies on both ends and matching traffic occurs on both concurrently, then absolutely.

And just to clarify. This is no issue at all. Two IKE_SAs or one IKE_SA won't make a difference in this scenario.

**Reply by Jiawei0227** [reply]
Thanks for your response! I really appreciate your help!

> I don't know that you mean here, because as soon as you load the connection (which you need to respond to the peer) that has a child section with start_action=trap, traffic can trigger an IKE_SA to get initiated on each of the peers concurrently. And you see that in the status output (the * behind the SPIs), each IKE_SA is initiated by one of them.

> The uniqueness check won't help if the SAs are created pretty much concurrently.

What I mean is that I have tried to set up the config like this to avoid race condition where both ends starts to send IKA_SA_INIT at the same time:

```
host1 
send
   trap
receive
   none

host2
send 
   none
receive
   none
```

Then after both side restart strongswan, there is only one IKE SA but there is only one CHILD_SA pair which is from `host1 send -> host2 receive`. Then I tried to manually change the host2 side config to `trap` as well and run `swanctl --load-conns`. I checked the log and it seems immediately the host2 starts to send `IKE_SA_INIT` to host1 to starts a new IKE SA, even though there is already one IKE SA existed. 

From https://wiki.strongswan.org/issues/2958  you mentioned that 

> However, when that responder then starts sending traffic to the initiator itself, it won't match the existing policy but the trap installed with transport-mode-out. As explained above, the existing IKE_SA created by the peer won't be reused, instead a new one is initiated. And because this new IKE_SA uses the same identities (the IPs) and is initiated from the same source IP and port the existing IKE_SA uses, it results in the following on the responder:

I was under the impression here the same thing would happen. Basically the existing IKE_SA will get replaced or recreated. But what I ended up having is actually two IKA_SA co-exist. So my question is whats the criteria here that the host2 decides to not reuse/replace the existing IKA_SA but to create a new IKA_SA?

**Reply by tobiasbrunner** [reply]
> Then I tried to manually change the host2 side config to `trap` as well and run `swanctl --load-conns`.

Note that you could install trap policies for a connection via `swanctl --install` command.

>  I checked the log and it seems immediately the host2 starts to send `IKE_SA_INIT` to host1 to starts a new IKE SA, even though there is already one IKE SA existed.

That's weird. I assume you only changed `start_action`, everything else stayed the same? (If the latter is true, you should see `updated vici connection: node2node` in the log, if not, you'd get `replaced vici connection: node2node`.)  Do you perhaps have the option `charon.reuse_ikesa` disabled? You can try increasing the log level for _mgr_ to 2 in order to see slightly more details about how IKE_SAs are checked out and created.


> So my question is whats the criteria here that the host2 decides to not reuse/replace the existing IKA_SA but to create a new IKA_SA?

The IKE and peer configs have to be exactly the same for an existing IKE_SA to get reused.
